core/net/rpl/rpl-ns.c

```diff
@@ -0,0 +1,211 @@
+node_matches_address(const rpl_dag_t *dag, const rpl_ns_node_t *node, const uip_ipaddr_t *addr)
+{
+  return addr != NULL
+      && node != NULL
+      && dag != NULL
+      && dag == node->dag
+      && !memcmp(addr, &node->dag->dag_id, 8)
+      && !memcmp(((const unsigned char *)addr) + 8, node->link_identifier, 8);
+}
+/*---------------------------------------------------------------------------*/
+rpl_ns_node_t *
+rpl_ns_get_node(const rpl_dag_t *dag, const uip_ipaddr_t *addr)
+{
+  rpl_ns_node_t *l;
+  for(l = list_head(nodelist); l != NULL; l = list_item_next(l)) {
+    /* Compare prefix and node identifier */
+    if(node_matches_address(dag, l, addr)) {
+      return l;
+    }
+  }
+  return NULL;
+}
+/*---------------------------------------------------------------------------*/
+int
+rpl_ns_is_node_reachable(const rpl_dag_t *dag, const uip_ipaddr_t *addr)
+{
+  int max_depth = RPL_NS_LINK_NUM;
+  rpl_ns_node_t *node = rpl_ns_get_node(dag, addr);
+  rpl_ns_node_t *root_node = rpl_ns_get_node(dag, dag != NULL ? &dag->dag_id : NULL);
+  while(node != NULL && node != root_node && max_depth > 0) {
+    node = node->parent;
+    max_depth--;
+  }
+  return node != NULL && node == root_node;
+}
+/*---------------------------------------------------------------------------*/
+void
+rpl_ns_expire_parent(rpl_dag_t *dag, const uip_ipaddr_t *child, const uip_ipaddr_t *parent)
+{
+  rpl_ns_node_t *l = rpl_ns_get_node(dag, child);
+  /* Check if parent matches */
+  if(l != NULL && node_matches_address(dag, l->parent, parent)) {
+    l->lifetime = RPL_NOPATH_REMOVAL_DELAY;
+  }
+}
+/*---------------------------------------------------------------------------*/
+rpl_ns_node_t *
+rpl_ns_update_node(rpl_dag_t *dag, const uip_ipaddr_t *child, const uip_ipaddr_t *parent, uint32_t lifetime)
+{
+  rpl_ns_node_t *child_node = rpl_ns_get_node(dag, child);
+  rpl_ns_node_t *parent_node = rpl_ns_get_node(dag, parent);
+
+  if(parent != NULL) {
+    /* No node for the parent, add one with infinite lifetime */
+    if(parent_node == NULL) {
+      parent_node = rpl_ns_update_node(dag, parent, NULL, 0xffffffff);
+      if(parent_node == NULL) {
+        return NULL;
+      }
+    }
+  }
+
+  /* No node for this child, add one */
+  if(child_node == NULL) {
+    child_node = memb_alloc(&nodememb);
+    /* No space left, abort */
+    if(child_node == NULL) {
+      return NULL;
+    }
+    list_add(nodelist, child_node);
+    num_nodes++;
+  }
+
+  /* Initialize node */
+  child_node->dag = dag;
+  child_node->lifetime = lifetime;
+  child_node->parent = parent_node;
+  memcpy(child_node->link_identifier, ((const unsigned char *)child) + 8, 8);
+
+  return child_node;
+}
+/*---------------------------------------------------------------------------*/
+void
+rpl_ns_init()
+{
+  num_nodes = 0;
+  memb_init(&nodememb);
+  list_init(nodelist);
+}
+/*---------------------------------------------------------------------------*/
+rpl_ns_node_t *rpl_ns_node_head()
```

Comment:
Missing line break between return type and function name here and in the following function.